### PR TITLE
DE28906: Fix for nested drop zones with rejected inner drop

### PR DESCRIPTION
**Pull request by object88**
#### What does this PR do?

Allows an outer nested drop element to receive a drop if an inner nested drop has rejected it.  In a nested configuration, an outer nest may receive a drop notification if an inner one doesn't handle it.  Once a drop is handled, the outward event propagation is stopped.

#### Any background context you want to provide?

When testing an upgrade in Churro, bumping react-dnd to the current version (0.9.6-rally.9), the team planning page broke in alm-guitests for workspaces that are configured with manual ranking.  If you drag something form the backlong onto the iteration's stories, the drop isn't allowed -- you must pass over the header row area, which is a functional regression.

#### Where should the reviewer start?

NOTE: The only two files that really changed here are: `modules/backends/Mouse.js` and `modules/utils/createDragDropMixin.js`.
The rest of the files are built based on that.

#### How should this be manually tested?

There are no tests for this package.

#### What are the relevant tickets?

[DE28906](https://rally1.rallydev.com/#/58644193630d/detail/defect/63060271808)

#### Screenshots, if appropriate. [:link:](http://recordit.co/)

n/a

#### Definition of Done
* [X] Story __acceptance criteria__ updated if changed
* [X] Does not negatively impacting [__perceived performance__](https://docs.google.com/document/d/1j4AbGqGp_Le1SfQp3lp-pclk0ojQqZy7Cp-HbNq4i94)
* [x] Does not degrade __performance__ and meets feature performance requirements
* [x] Renders properly in [__all supported browsers__](https://help.rallydev.com/supported-web-browsers)
* [X] __Sufficiently rich data__ used during development to expose important data relationships
* [x] __Code reviewed__ by someone who didn't write it
* [x] __Automated tests__ written and passing at appropriate levels. (unit, integration, end-to-end, etc.)
* [x] No functional __regressions__ have been introduced

### CODE REVIEW

**Comment by brentertz** [comment]
Though there are no tests in this package, there is a demo that contains some different use cases.  It is useful to ensure that all of the examples still operate properly.

**Comment by brentertz** [comment]
If this will cause a regression to the sombrero-demo, then it should be addressed as part of this story.

**Comment by object88** [comment]
@brentertz , I see the Sombrero demos, but it's not always well documented, what the exact expected behaviors are.

We do want a drop on a nested drop target to trigger a drop on the parent drop target; there is a test in alm-guitests which is this scenario.  The linked defect goes into some more detail.

**Comment by object88** [comment]
Perhaps I should say... for the purposes of the team planning page in manual ranking mode, which shows a defect when using react-dnd 0.9.6-rally.9, we want the drop to trickle through to the outer nested drop.  When we are in manual ranking, the Sombrero DataTable that renders the stories in the iteration view disables drag & drop.  In 0.9.6-rally.6, the logic somehow got the drop event to the outer drop, so we could drop to the table as a whole.  With the current version, the drop goes to the inner target, and halts there -- the only way to recognize the table as a drop target was to drag over the header row first.

If this behavior is bad for other parts of our app, I haven't seen a test yet which breaks it.

File: dist-modules/backends/Mouse.js [hunk 1]
```diff
@@ -107,9 +96,9 @@ function handleDragStart(component, type, e) {
 }
 
 function handleTopMouseUp(e) {
-  if (_currentDropTargetComponent) {
-    _currentDropTargetComponent.handleDrop(getDragItemTypes(), e);
-  }
+  _currentDropTargetComponents.forEach(function (cdtc) {
+    return cdtc.handleDrop(getDragItemTypes(), e);
```
Comment (limscoder) [inline dist-modules/backends/Mouse.js]: The PR description says the outer container's handler is only called when the inner container does not handle the drop, but this code looks like all containers handle the drop.

Which is it supposed to do?

Comment (object88) [inline dist-modules/backends/Mouse.js]: My apologies -- I think that I wasn't clear at first.  All of the nested drop targets will receive notification, regardless of how the inner-nested targets handle the drop.

Comment (limscoder) [inline dist-modules/backends/Mouse.js]: I'd be more comfortable if the code walked inner->outer and stopped at the first successful callback. That would preserve existing behavior while still working with your use-case (I think), and not leading to future confusion around multiple callbacks being fired for the same drop.

Comment (object88) [inline dist-modules/backends/Mouse.js]: I'm following up on this now.  Are you only interested in this for the handleDrop?  I feel like it's OK to go through the whole nested stack for `leave`, `over`, and `enter`, as in the above method.

**Comment by object88** [comment]
Do not merge yet: we are not correctly handling sort order for nested drop targets.

File: dist-modules/backends/Mouse.js [hunk 2]
```diff
@@ -107,9 +102,9 @@ function handleDragStart(component, type, e) {
 }
 
 function handleTopMouseUp(e) {
-  if (_currentDropTargetComponent) {
-    _currentDropTargetComponent.handleDrop(getDragItemTypes(), e);
-  }
+  _currentDropTargetComponents.some(function (cdtc) {
```
Comment (limscoder) [inline dist-modules/backends/Mouse.js]: Is this array always ordered inner->outer?

File: dist-modules/backends/Mouse.js [hunk 3]
```diff
@@ -1,18 +1,19 @@
 "use strict";
 
-var ReactDOM = require("react-dom"),
+var _ = require("lodash"),
```
Comment (limscoder) [inline dist-modules/backends/Mouse.js]: Can we require only the lodash functions that are needed?

`difference = require("lodash/array/difference")`?

Comment (object88) [inline dist-modules/backends/Mouse.js]: Can do!

File: dist-modules/backends/Mouse.js (hunk 2 above)
Comment (object88) [inline dist-modules/backends/Mouse.js]: Great question -- I just got caught by that going over the Sombrero demo earlier.  I added `return matchingDropTargets.sort` to the `findDropTargets` to handle this; should be working now.

**Comment by limscoder** [comment]
Only thing I'm still a bit concerned about is adding more logic to these handlers that get called on every mouse mouse event. Might want to double check that it hasn't introduced any extra lag on crappy hardware. Other than that it looks good.

**Comment by object88** [comment]
I will check on it.  I feel that it shouldn't be too much worse -- thankfully this is only applied during a drag, and the number of drop targets was already reduced (in the original code) to just what's under the mouse.  In the most heinous situations that I've seen... that's 2.  Will update after doing some comparisons.

**Comment by object88** [comment]
I used the Team Plan page to test dnd from the backlog to an iteration.  The Chrome tools are extremely helpful, in that they tell me what part of the timeline is spent dragging.  (Thanks, Chrome!)

react-dnd 0.9.6-rally.9: almost entirely in the 50-60fps range, with occasional dipping to 40-45fps.

react-dnd 0.9.6-rally.next: almost entirely in the 50-60fls range, with occasional dipping into 30-45fps.

**Comment by limscoder** [comment]
@object88 looks good.

One small thing: the .sort could use a static function instead of a closure.

**Comment by object88** [comment]
Will get that in.  Thanks again, @limscoder!

**Pull request merged**